Voting dapp tutorial (#1)
raduciobanu22/docs: docs/build/voting-dapp-tutorial.md

---
title: dApp Tutorial
description: Build a complete dApp on Kadena
---

import PageRef from '@components/PageRef'
import Tabs from '@theme/Tabs';
import TabItem from '@theme/TabItem';

# Building a voting dApp

In its most basic form, Kadena blockchain is a digital ledger. This ledger is never stored, but rather exists on the “chain” supported by thousands of nodes simultaneously. Thanks to encryption and decentralization, Kadena blockchain’s database of transactions is incorruptible, and each record is easily verifiable. The network cannot be taken down or influenced by a single party because it doesn’t exist in one place.

> Blockchain can solve the many problems discovered in these early attempts at online voting. A blockchain-based voting application does not concern itself with the security of its Internet connection, because any hacker with access to the terminal will not be able to affect other nodes. Voters can effectively submit their vote without revealing their identity or political preferences to the public. Officials can count votes with absolute certainty, knowing that each ID can be attributed to one vote, no fakes can be created, and that tampering is impossible.

Reviewer: I like that the introduction helps motivate why blockchain / digital ledger technologies are a good fit for remote voting. I think this section can be made a little stronger, though, by pointing at some real-world blockchain voting success and by being more specific about how blockchain technologies make online voting more secure. For example:

> Blockchain can solve the many problems discovered in these early attempts at online voting.

If we assert that blockchain technology solves problems with earlier methods of online voting, then I think we should back it up by explaining some specific advantages of blockchain technology. There are some fairly sophisticated schemes for non-blockchain online voting that security researchers have put together, some of which have seen real-world use (such as [Helios](https://vote.heliosvoting.org/faq)) and so I think this can come off as a little flippant without backup.

For example, there are signifiant issues with existing blockchain voting solutions like the Voatz app used in the West Virginia elections in 2018, [as described in this security analysis](https://www.usenix.org/system/files/sec20-specter.pdf). Ultimately election officials disregarded the Voatz-provided votes. Another great piece is [Going from bad to worse: from Internet voting to blockchain voting](https://static1.squarespace.com/static/59aae5e9a803bb10bedeb03e/t/615c75f9eaae697e540f5c4f/1633449465655/VotingPaper-Rivest%2C+Narula%2C+Sunoo.pdf).

Fortunately, the security issues have nothing to do with the blockchain itself — it's just that there's a lot more to election security than the parts replaced by blockchain technologies. So what does blockchain bring to the table?

[Efficient, Coercion-free and Universally Verifiable Blockchain-based Voting](https://eprint.iacr.org/2019/1406.pdf) goes into some detail on this subject, but the biggest takeaways for me:

* The blockchain's append-only structure makes it easy to verify votes without trusting a central election administrator to tally the votes and report them accurately.
* Kadena's rapid transaction speed means that this method scales to large votes, as compared to say paper voting, which is enormously expensive.
* It is easier to provide voters the ability to verify their vote was tallied correctly, if you are willing to give up on receipt-freeness (ie. the restriction that voters should not be able to prove how they voted, which is used to prevent vote buying in critical elections).

All of this said, it might be easier to just hand-wave all this and say that voting is an interesting application of blockchain technology and point at some real-world successes:

* Thailand's Democrat party held an election in which [more than 120,000 registered Democrats voted via blockchain](https://bitcoinmagazine.com/culture/thailand-uses-blockchain-supported-electronic-voting-system-primaries).
* Switzerland ran a small-scale blockchain vote, which [they hailed as a success](https://www.swissinfo.ch/eng/business/crypto-valley-_-switzerland-s-first-municipal-blockchain-vote-hailed-a-success/44230928).

Author: @thomashoneyman

> In its most basic form, Kadena blockchain is a digital ledger. This ledger is never stored, but rather exists on the “chain” supported by thousands of nodes simultaneously. Thanks to encryption and decentralization, Kadena blockchain’s database of transactions is incorruptible, and each record is easily verifiable. The network cannot be taken down or influenced by a single party because it doesn’t exist in one place.
> 
> Blockchain can solve the many problems discovered in these early attempts at online voting. A blockchain-based voting application does not concern itself with the security of its Internet connection, because any hacker with access to the terminal will not be able to affect other nodes. Voters can effectively submit their vote without revealing their identity or political preferences to the public. Officials can count votes with absolute certainty, knowing that each ID can be attributed to one vote, no fakes can be created, and that tampering is impossible.

Reviewer: ```suggestion
One of the best ways to learn a new technology is to get your hands dirty and build applications with it. In this tutorial we'll learn to use Kadena's smart contract language, Pact, to build a minimal voting application that runs on the Chainweb blockchain.

Along the way, we'll learn how to write and test smart contracts in Pact, the basic architecture of dApps on Chainweb, and about major concepts like gas stations and deploying contracts. We'll also get acquainted with several useful tools in the Kadena ecosystem, including the Pact local test server and JavaScript libraries you can use to interact with nodes running Pact.

**Voting on the Blockchain**

Elections are a necessary part of democracies and democratic organizations. The voting systems used to administer elections must ensure a fair process and trustworthy result — easier said than done! Election security is a deep, fascinating topic, especially when it comes to online voting.

Blockchain technologies are well-suited to help secure online elections. A public blockchain gives participants a single view of all transactions, which makes it easy to verify votes without trusting a central election authority to tally and report the results. Kadena's rapid transaction speeds (480,000 per second, when using 20 chains) scale voting to even large elections.

Blockchain technologies don't solve all election security issues, but they're a strong foundation and [researchers have proposed fully-secure online voting systems based on them](https://eprint.iacr.org/2019/1406.pdf). They've also seen success in the real world: Thailand's Democrat party held an election in which [more than 120,000 registered Democrats voted via blockchain](https://bitcoinmagazine.com/culture/thailand-uses-blockchain-supported-electronic-voting-system-primaries).

**What We're Building**

We'll build a tiny voting dApp prototype that lets anyone with a Kadena wallet address vote for a candidate from a selection of candidates. Each voter (ie. address) can vote once. Some Kadena accounts are chosen as "election officials", and the smart contract grants them special privileges to select the candidates. Election officials can add new candidates at any time (but they can't remove candidates or adjust their votes).

Once the app is deployed, the election has begun! The frontend for our dapp will help users submit their votes and will display the total votes received by each candidate.
```

---

## Overview

> In this tutorial we will be building a voting application on the Kadena blockchain. The application will consist of three distinct layers:

Reviewer: As you read the tutorial it becomes clear that the voting app is just a vehicle to teach general Kadena / Pact / Chainweb development concepts, and that the tutorial is targeted at a beginner (someone who hasn't even installed Pact yet). 

I think the introduction should make this point clear up front. For example:

> One of the best ways to learn a new technology is to get your hands dirty and build applications with it. In this tutorial we'll learn to use Kadena's smart contract language, Pact, to build a minimal voting application that runs on the Chainweb blockchain.

I think it's helpful to be concrete about what a reader will take away from the tutorial, so they can judge up front whether the tutorial is the right fit for them. After all, we're asking for potentially hours of their time to go through this tutorial and learn with us! For example:

> Along the way, we'll learn how to write and test smart contracts in Pact, the basic architecture of dApps on Chainweb, and about major concepts like gas stations and deploying contracts. We'll also get acquainted with several useful tools in the Kadena ecosystem, including the Pact local test server and JavaScript libraries you can use to interact with nodes running Pact.

Since this is a lengthy tutorial I also think it makes sense to give readers an idea of what's to come. I like that you describe the three distinct layers in the app (smart contract backend, web app frontend, and gas station), but this is an architecture overview — what's missing is a quick overview of how the application will work at a high level. For example:

> We're building a tiny voting application that lets anyone with a Kadena wallet address vote for a candidate from a selection of candidates. Each voter (ie. address) can vote once. Some Kadena accounts are chosen as "election officials", and the smart contract grants them special privileges to select the candidates. Election officials can add new candidates at any time (but they can't remove candidates or adjust their votes).
> Once the app is deployed, the election has begun! The frontend for our dapp will help users submit their votes and will display the total votes received by each candidate.

@LindaOrtega's note about a table of contents is also good, and I agree that would be useful.

It's only after all this setup that I think the architectural overview makes sense, because at that point we're ready to tie the features we've described into the actual implementation.

Author: @thomashoneyman

> 1. Smart contract (back-end)
> 2. Web app (front-end)
> 3. Gas Station

Reviewer: This is a suggestion for a table of contents with anchor links deeper into the tutorial so readers can skip around; feel free to change this if you shake up the structure of the piece.

```suggestion
**Table of Contents**

1. Building the Smart Contract Backend
2. Testing our Smart Contract
3. Implementing a Gas Station
4. Building the Frontend Voting App
5. Deploying to Chainweb
```

> ## Environment Setup

Reviewer: I think this whole section should be removed. I don't think tutorials should tell readers how to install Pact (or other Kadena technologies). There should be an authoritative document with installation instructions, and all tutorials should link to it. This principle holds true for any common Kadena workflows (installing Atom extensions, detailed steps for deploying to testnet, etc.) and certainly for instructions that are likely to change in the future.

This helps ensure these resources stay up-to-date and it increases the pressure to make the authoritative installation instructions really good.

Author: @raduciobanu22

### Install Pact

#### Mac

  ```bash
  brew update
  brew install kadena-io/pact/pact
  ```

#### Linux
  * Using **nix**

  To build with **Nix** follow the setup instructions [here](https://github.com/kadena-io/pact/wiki/Building-Kadena-Projects).
  Once the build is finished, you can run Pact with the following command:

  `./result/ghc/pact/bin/pact`

  For other options to build from source please refer to [this section](https://github.com/kadena-io/pact#building-from-source).

**Verifying Installation**

Type `pact` in a terminal and try out some commands:

```clojure
$ pact
pact> (+ 1 2)
3
pact> (+ "hello, " "world")
"hello, world"
pact> (< 0 1)
true
```

### Create Project Structure

```bash
mkdir election-dapp && cd election-dapp
mkdir pact
mkdir src

> ```

Reviewer: This is the only part of this section that I think should remain. I think it's also worth writing a sentence or two about why there are `pact` and `src` directories — what are we going to put in each one?

Author: @raduciobanu22

### Atom IDE (Optional)

You can use any text editor to write Pact but if you prefer the benefits of an IDE, "language-pact" is a package for [Atom](https://atom.io) that provides syntax highlighting and linting.

To install it, go to `Preferences -> Packages -> Type "language-pact"` and click the `Install` button.

![atom-language-pact](/img/docs/voting-dapp/atom-language-pact.png)

> ## Smart Contracts

Reviewer: I might rename this section "Implementing the Voting Smart Contract" or something like that to help the reader see that in this section we translate our dapp's features into smart contract code (we're writing a voting smart contract).

Author: @raduciobanu22

Our voting app will allow you to submit a vote while preventing an address to vote more than once.

> Additionally we will use a *gas station* to fund the gas fees for interacting with our contract, meaning our users don't have to worry about paying for gas.

Reviewer: This feels a little early to be introducing the gas station concept. We don't get around to using it until line ~431. Since this tutorial is for beginners, I think we should try to introduce as few concepts at a time as possible. This could be one of their first smart contracts they're seeing with Pact! For that reason, I think we should keep this section strictly about the smart contract itself and push all the gas station discussion until we actually implement it.

Author: @raduciobanu22

:::info
**Gas** is the cost necessary to perform a transaction on the network. Gas is paid to miners and its price varies based on supply and demand. It's a critical piece of the puzzle, but at the same time it brings up a UX problem. Every user needs to be aware of what gas is as well as how much gas they need to pay for their transaction. This causes significant friction and a less than ideal experience.

To help mitigate this problem Kadena brings an innovation to the game. Hello [gas stations](https://medium.com/kadena-io/the-first-crypto-gas-station-is-now-on-kadenas-blockchain-6dc43b4b3836)!

Gas stations are a way for dApps to subsidize gas costs for their users. This means that your user doesn't need to know what gas is or how much the gas price is, which translates into a smooth experience when interacting with your dApp.
:::

### Voting

A typical developer workflow looks like this:

1. Write contract code in `.pact` files
2. Write tests in `.repl` files
3. Execute your tests in the repl
4. Deploy to local pact server

> 5. Deploy to testnet

Reviewer: 👍 It's great to provide a mental scaffold for how to approach development.

In this section we will be focusing on steps 1-3 and later deploy to local pact server and testnet.

Assuming you are still in your project directory, let's create the `election.pact` and `election.repl` files:

```clojure
touch pact/election.pact pact/election.repl
```

> Remember, `election.pact` is the contract source code while in `election.repl` we'll write tests.

Reviewer: Perhaps it's worth a sentence or two about the Pact REPL, noting that you can load Pact contracts and interactively work with them, and link over to some more in-depth documentation about the REPL for interested readers to follow?

Author: @thomashoneyman

> In this section we will be focusing on steps 1-3 and later deploy to local pact server and testnet.
> 
> Assuming you are still in your project directory, let's create the `election.pact` and `election.repl` files:
> 
> ```clojure
> touch pact/election.pact pact/election.repl
> ```
> 
> Remember, `election.pact` is the contract source code while in `election.repl` we'll write tests.

Reviewer: ```suggestion
In this section we will focus on steps 1 to 3. Later, we'll deploy our smart contract to a local Pact server and to testnet (the test version of Chainweb).

In your project directory, let's create two files:

* `pact/election.pact`, which will hold the source code for our smart contract
* `pact/election.repl`, which will hold our tests

The Pact REPL is an environment where we can load our Pact source code and work with it interactively. It's a best practice to include a `.repl` file next to your source code which imports your contract, calls functions from it, and inspects its current state to ensure everything is correct.
```

> We will also need the `coin` contract which requires the `fungible-v2` interface for our tests. You can get the latest `coin-v3` [here](https://github.com/kadena-io/chainweb-node/blob/master/pact/coin-contract/v3/coin-v3.pact) and `fungible-v2` [here](https://github.com/kadena-io/chainweb-node/blob/master/pact/coin-contract/v2/fungible-v2.pact). Add these to your project in the following directory: `pact/root/`.

Reviewer: As a beginner, I feel like this comes out of nowhere. What is the `coin` contract and why do we require it? What is `fungible-v2`? If I'm developing my own project, how would I know that I need these contracts? Is it common practice to copy files from the Kadena GitHub into my own source code, or are we just doing that for the tutorial? What happens when `fungible-v3` and `coin-v4` come out, will my contracts break?

This also introduces a new architectural concept: coping the files into `pact/root`. Is this a typical directory structure when pulling in "library" files like `coin-v3`?

Author: > This also introduces a new architectural concept: coping the files into `pact/root`. Is this a typical directory structure when pulling in "library" files like `coin-v3`?

Yes, I used Marmalade as template for the structure.

> As a beginner, I feel like this comes out of nowhere. What is the coin contract and why do we require it? What is fungible-v2? If I'm developing my own project, how would I know that I need these contracts? Is it common practice to copy files from the Kadena GitHub into my own source code, or are we just doing that for the tutorial? What happens when fungible-v3 and coin-v4 come out, will my contracts break?

I agree that this requires more explanation.

Author: I provided additional explanation.

> We are now set to start implementing our contract, so let's copy the following code in the `election.pact` file:

Reviewer: I think this is a good place to pause and recap the features we're about to implement. Namely:

1. Voters can record 1 vote for a candidate of their choice from the list of options. In response to voting, we'll return confirmation of their vote by returning their vote back to them.
2. Election administrators can add candidates (so the "add candidate" functionality should be guarded to only allow access to keys belonging to the election administrator accounts).
3. And so on...

This helps the reader remember what the features of the app are, and therefore why we're adding a `'election-admin-keyset`, a `GOVERNANCE` capability, and so on. I think we should keep referring back to the features we're implementing so that it's clear to the reader _why_ we're writing the contract this way.

Reviewer: I think this implementation section could be broken down into sub-sections, as a lot of concepts get introduced. Some sections that come to mind:

* Voting contract module
* The governance capability
* Tables and data storage
* Testing the contract

Breaking it down into sub-sections would help you write a little intro section for each one, in which you can describe what the feature is for and how we'll use it to implement our dapp, and then conclude with a little summary section that recaps what you did. This kind of intro -> body -> recap structure works really well for helping readers absorb the material.

> ```clojure

Reviewer: It can be helpful to include a comment at the top of code snippets that help readers see what file it comes from (even though you already noted this is `election.pact`). For example:

```suggestion
```clojure
;; election.pact
```

Author: @raduciobanu22

(define-keyset 'election-admin-keyset)

(module election GOVERNANCE
  "Election demo module"

  (defcap GOVERNANCE ()
    "Only admin can update this module"
    (enforce-keyset 'election-admin-keyset))

  (defun vote (key:string)
    "Submit a new vote"
    (format "Voted {}!" [key]))
)

(vote "A")
```

> In this snippet we defined a module named `election` that includes a `vote` function which simply displays its parameter value as part of a formatter string.

Reviewer: It would be useful to link to the Pact language reference on format strings.

Author: @raduciobanu22

> You can also notice the `GOVERNANCE` keyword next to the module name and the capability with the same name defined below. It is called *module governance capability* and it can be as simple as our example, enforcing a keyset or more complex like tallying a stakeholder vote on an upgrade hash.

Reviewer: As a beginner, it's not clear to me what the governance capability is actually providing here. What would be different about this contract if this capability were not specified? We should spell out concretely, with respect to the voting app we're writing, what this feature provides us.

Right now all we have is the comment "Only admin can update this module" and the note that this capability can enforce a keyset, but it's not clear what that really means. Later on you demonstrate guarding a function within the module with this capability, so maybe we can just add a sentence here that says something like:

> The `GOVERNANCE` capability can be used to restrict who can call particular functions in this module. For example, later on we'll add an `insert-candidate` function that only election administrators should be able to call, and we'll guard it with the `GOVERNANCE` capability.

:::note
Module names and keyset definitions are required to be unique. We will mention this again when we get to deploy our contract to testnet.
:::

We've got some code written, now let's test it. Copy the code below in the `election.repl` file:

```clojure

> ;; begin a transaction

Reviewer: I like the inline comments. They do a good job helping describe what's going on one line at a time.

However, I feel like each of these comments could use one more sentence that describes _why_ it's necessary. Here are some questions that come to mind as a beginner reading this, which we might want to answer:

* Why does the repl file have to begin with `(begin-tx)` and end with `(commit-tx)`? What would happen if that weren't there?
* Whe do we have to add `["admin-key"]` to the environment? Where did "admin-key" come from? Is it just an arbitrary string? Could I add a bunch of random keys here? What would that mean?
* What does "pred": "keys-all" mean wrt the keyset?

You don't necessarily need to explain all this stuff. But you could add links to Kadena docs that do explain it, or add a sentence summarizing, just so that readers have a general sense of why these lines were written, so they can adapt it for when they need to do something similar in their own projects.

(begin-tx)
;; set transaction signature key to my-key
(env-keys ["admin-key"])

;; Add the election-admin-keyset to environment data
(env-data {
  'election-admin-keyset: { "keys": ["admin-key"], "pred": "keys-all" }
})
;; load election.pact into the REPL
(load "election.pact")
;; commit the transaction
(commit-tx)
```

:::info

> `begin-tx`, `env-keys`, `env-data` are "repl-only" functions, they are automatically loaded in the REPL to help simulate blockchain environment but they are not available for blockchain-based execution. You can find the complete "REPL-only" functions list [here](https://pact-language.readthedocs.io/en/latest/pact-functions.html?highlight=repl%20only#repl-only-functions).

Reviewer: > they are automatically loaded in the REPL to help simulate blockchain environment but they are not available for blockchain-based execution.

I think it's useful to note what the blockchain equivalents are for these repl-only functions. Presumably `begin-tx` corresponds with actually submitting a transaction to peers, `env-keys` corresponds with the keys used to sign that transaction, and `env-data` is an arbitrary JSON payload?

I see that in the next info box you clarify this at least for `env-data`, but I feel like these info boxes could be merged if you described how each of the repl-only functions maps to the blockchain equivalents.

:::

:::info
Each Pact transaction is submitted with raw JSON data that can be read by functions like `read-msg`. In REPL scripts we are using `env-data` to attach any data to this JSON, including keysets which are then read with `read-keyset`.
:::

Run the script in your terminal:
```clojure
$ pact
pact> (load "election.repl")
```
The output should be similar to the one below:
```clojure
"Loading election.repl..."
"Begin Tx 0"
"Setting transaction keys"
"Setting transaction data"
"Loading election.pact..."
"Keyset defined"
"Loaded module election, hash jOBJiSEH5HVsAmLNSdHq_D3Kl6TkFpCJcEWPtJD86Gc"
"Voted A!"
"Commit Tx 0"
```

> 

Reviewer: It would be nice to add a little wrap-up summary of what we've accomplished so far before transitioning into the next section. This feels a little jarring to have run the repl file and immediately be jumping into tables without a motivation for what we've done and why we're moving on.

For example, something like: We've written a bare-bones module and an accompanying test repl file that can be used to execute it in a simulated blockchain environment — and now, because we need to store some dynamic data (who the candidates are, so you can vote for them, as well as who voted, so we can disallow multiple votes), we'll move on to tables

Pact smart contracts store data in tables so let's define the ones that we need for our voting contract. Copy the code below after the governance capability definition:

```clojure
  (defschema candidates-schema
    "Candidates table schema"
    name:string

>     votes:integer)

Reviewer: I think this file could also use the inline comments you've been doing. For example, you could briefly state that these are our two columns, and you can specify a type for them (and maybe link to the types that Pact supports)

Author: @raduciobanu22

  (defschema votes-schema
    "Votes table schema"

>     cid:string

Reviewer: What is `cid`? Can we use a more informative variable name?

Author: @raduciobanu22

  )

  (deftable votes:{votes-schema})

  (deftable candidates:{candidates-schema})
  ```

> What we did here:

Reviewer: I feel like this manual description of what we did could be turned into inline comments, and instead we can summarize at a higher level what we did: we created two tables: one for storing candidates and their associated vote counts, and one for storing what accounts have already voted (so we can prevent double-voting).

It would be useful to link to Kadena docs on how exactly these tables work on the blockchain (how and where are they stored?) as well.

* `defschema candidates-schema` -> defines the schema for our **candidates** table where we will keep track of the candidates that users can vote for
* `defschema votes-schema` -> defines the schema for our **votes** table where we will store each vote
* `deftable votes:{votes-schema}` -> defines the `votes` table that will use the schema defined above
* `deftable candidates:{candidates-schema}` -> defines the `candidates` table that will use the schema defined above

:::note

> Pact implements a key-row model which means a row is accessed by a single key. It is our responsibility as developers to design the schema in a way that we can retrieve the information that we need using a single row query. Multiple row queries are very expensive and shoud not be used.

Reviewer: Does this mean that a table schema always has to begin with a string column which serves as the key?

Author: Yes, the key is a string and it's not explicitly defined in the schema. I admit I was confused initially and I added it in the schema definition.

The row key is always a simple string, to not be confused with the cryptographic keys used for signing transaction.
:::

Since we have defined our data storage, we can now implement the `vote` function:

> ```clojure

Reviewer: Are we still in the `election.pact` file? Is the `use` statement below supposed to be at the top of the file?

>   ; import coin.details function

Reviewer: As with some of the the other inline comments — it's helpful to explain that we're importing the coin.details function, but what I really need to know is _why_ we're importing this function and what it can be used for, ideally with a permalink to the implementation or some docs.

Later in the code I can see that that it's being applied to the account, but I don't know what the result of applying this function to the account string does. That's something we should put in an inline comment.

  (use coin [ details ])

  (defcap ACCOUNT-OWNER (account:string)
    "Make sure the requester owns the KDA account"
    (enforce-guard (at 'guard (coin.details account)))

>   )

Reviewer: I feel like this needs a couple lines of explanation inline. We seem to be defining a capability, ACCOUNT-OWNER, that takes in an account ID and...ensures that the person who signed this transaction is the owner of their own account? I'm not totally following what this is doing.

  (defun user-voted:bool (account:string)
    "Check if a user already voted"

    (with-default-read votes account
      { "cid": "" }
      { "cid":= cid }

>       (> (length cid) 0))

Reviewer: This is our first example of reading from a table; this also deserves a couple lines of explanation. What is `with-default-read`? Are there non-default reads? What are these five arguments supposed to mean?

  )

  (defun candidate-exists:bool (cid:string)
    "Check if a candidate exists"

    (with-default-read candidates cid
      { "name": "" }
      { "name" := name }
      (> (length name) 0))

  (defun vote-protected (account:string candidateId:string)
    (require-capability (ACCOUNT-OWNER account))

    (with-read candidates candidateId { "votes" := votesCount }
      (update candidates candidateId { "votes": (+ votesCount 1) })
      (insert votes account { "cid": candidateId })
      (emit-event (VOTED candidateId))
    )

>   )

Reviewer: This needs explanation, too.

What is `(require-capability ...)`? (It feels self-explanatory, but we could explain that this will bail out the execution of this function and fail if this verification check fails.)

Why are we using `with-read` here instead of `with-default-read`? What does it mean to assign with `:=`?

What does it mean to emit an event? Where did `VOTED` come from? Does the all-caps mean something special?

  (defun vote (account:string cid:string)
    "Submit a new vote"

    (let ((double-vote (user-voted account)))
      (enforce (= double-vote false) "Multiple voting not allowed"))

    (let ((exists (candidate-exists cid)))
      (enforce (= exists true) "Candidate doesn't exist"))

    (with-capability (ACCOUNT-OWNER account)
      (vote-protected account cid))

    (format "Voted for candidate {}!" [cid])
  )
```

> Quite a lot happening here but don't worry we'll explain everything:

Reviewer: I think I would find this easier to follow if this explanation largely moved into inline comments. For example:

> * while the `ACCOUNT-OWNER` capability is in scope we are calling `vote-protected` which is the function that updates the database. First it checks that we already have the capability (this means the function cannot be called directly) then proceed to update the votes of a candidate and record the vote in the votes table. At the end we emit the `VOTED` event which is useful for our frontend to update itself in real-time by listening to this event.

This would be easier for me to read if it were written into the function that it's describing as a comment. The bullet points below are all good, it's just that they would be easier to parse in the code.

:::tip
Spend a bit of time understanding the `vote` function implementation before moving forward.
:::

Now that we can record votes, we also need to read them:

```clojure
(defun get-votes:integer (cid:string)
  "Get the votes count by cid"
  (at 'votes (read candidates cid ['votes]))
)
```

Last thing on the list is adding candidates:

```clojure
(defun insert-candidate (candidate)
  "Insert a new candidate, admin operation"
  (with-capability (GOVERNANCE)
    (let ((name (at 'name candidate)))
    (insert candidates (at 'key candidate) { "name": (at 'name candidate), "votes": 0 })))
)

(defun insert-candidates (candidates:list)
  (map (insert-candidate) candidates)
)
```

Inserting a new candidate is an "admin-only" operation and we reused the already defined `GOVERNANCE` capability to enforce this.

With this we have esentially completed our module. All the required functionality is implemented.

When a module is deployed, the tables that it defines need to be created. This is done through `create-table` function. Insert the snippet below after the module's closing parenthesis:

```clojure
(if (read-msg "upgrade")

>   ["upgrade"]

Reviewer: Does the reader need to know what this is?

  [
    (create-table candidates)
    (create-table votes)
  ]
)
```

:::info
 Code outside the module will be called when the module is loaded the first time, when its deployed or upgraded. In the snippet above we are checking if the `upgrade` key that comes from transaction data is `true` and only execute the `create-table` calls if it's not since we cannot recreate tables when upgrading a module.
:::

:::tip
You can find the source code of the `election.pact` contract [here]. TODO: Insert repo link
:::

> We have written our contract but our work is not done yet. We need tests to validate that the functionality is correct. Clear the contents of the `election.repl` file and copy the following snippet:

Reviewer: Before we move on, I think there should be another summary section where you recap the major things you did / major features you used in this contract so that the reader can get a little help synthesizing all the new information they just got.

For example, you could recap that we used module capabilities to guard candidate creation to only election administrators and to guard voters so they can only vote from their own account (if I'm understanding that correctly). You could recap on how the tables were designed — earlier you mentioned how important single-row queries are, so you could restate that quickly and note that that's what we did. And you could mention that you can ensure particular code conditions are met with `enforce`.

This summary section is where you can help the reader by generalizing the concrete code you've written into takeaways on how to write Pact code in general.

```clojure

> ;; begin a transaction

Reviewer: Why is this first transaction that just loads environment data required? Could we explain that to the reader?

Reviewer: I think each of these code snippet sections should have a paragraph that tells the reader what we're about to do and why. 

For example, this code snippet could have a preceding paragraph that says we have to initialize the tables, set the admin keyset, and so on (if we really _do_ have to do that).

(begin-tx)
(env-data {
  ;; set environment data to the admin-keyset with keys admin-key and predicate function of keys-all
  'election-admin-keyset: { "keys": ["admin-key"], "pred": "keys-all" },
  'alice-keyset: { "keys": ["alice-key"], "pred": "keys-all" },
  'bob-keyset: { "keys": ["bob-key"], "pred": "keys-all" },
  'upgrade: false
})

;; load fungible-v2 interface required by coin module
(load "root/fungible-v2.pact")
;; load coin module
(load "root/coin-v3.pact")
;; load election module
(load "election.pact")
;; commit the transaction
(commit-tx)

(begin-tx)
;; create "alice" KDA account
(coin.create-account "alice" (read-keyset "alice-keyset"))
;; create "bob" KDA account
(coin.create-account "bob" (read-keyset "bob-keyset"))
(commit-tx)

(begin-tx)
;; import election module
(use election)

> (env-sigs [{ "key": "alice-key", "caps": []}])

Reviewer: What is `env-sigs`? What is `caps`?

(expect-failure "Can't vote for a non-existing candidate" (vote "alice" "5"))
(commit-tx)
```

> The comments should give you a fairly clear picture of what is happening here. We are using `expect-failure` to test if `vote` correctly fails if we try to vote for a candidate that does not exist.

Reviewer: ```suggestion
We are using `expect-failure` to test if `vote` correctly fails if we try to vote for a candidate that does not exist.
```

:::info
`expect` is one the functions that's only available in the REPL but there are many more, check them out <a href="https://pact-language.readthedocs.io/en/latest/pact-functions.html#repl-only-functions">here</a>.
:::

And more tests:

```clojure
(begin-tx)
(use election)
;; test if votes count for candidate "1" is initialized with 0
(expect "votes for Candidate A initialized" (get-votes "1") 0)
;; test if votes count for candidate "2" is initialized with 0
(expect "votes for Candidate B initialized" (get-votes "2") 0)
(commit-tx)

(begin-tx)
(use election)

> (env-sigs [{ "key": "alice-key", "caps": [(coin.GAS), (election.ACCOUNT-OWNER "alice")]}])

Reviewer: If you follow my earlier comment about delaying mention of the gas station, then I'd just add a comment here saying "After this code snippet we'll talk more about gas and gas stations, which are core Kadena concepts." I believe this is the first time it's shown up in the code.

;; test if votes count for candidate "1" is correctly increased by 1
(let ((count (get-votes "1")))
  (vote "alice" "1")
  (expect "votes count is increased by 1" (get-votes "1") (+ count 1)))

(expect "voted event"
  [ { "name": "election.VOTED", "params": ["1"], "module-hash": (at 'hash (describe-module "election"))}]
  (env-events true))

(env-sigs [{ "key": "bob-key", "caps": [(coin.GAS), (election.ACCOUNT-OWNER "bob")]}]}])
;; test if votes count for candidate "2" is correctly increased by 1
(let ((count (get-votes "2")))
  (vote "bob" "2")
  (expect "votes count is increased by 1" (get-votes "2") (+ count 1)))

(expect "voted event"
  [ { "name": "election.VOTED",
      "params": ["2"],
      "module-hash": (at 'hash (describe-module "election"))
    }
  ]
  (env-events true))

(expect-failure "Double voting not allowed" (vote "bob" "1"))

(commit-tx)
```

> Here we are testing that the `vote` function correctly increments the vote count and the `VOTED` event is emitted. We are also checking if we correctly handled the double voting scenario.

Reviewer: I think we should put this before the code snippet as an introduction. For example, something like:

> We've written some bare-bones tests that initialize our tables and ensure that you can't vote for a candidate that doesn't exist. But now we should write some more exhaustive tests that cover our dapp features: ...

:::tip
Can you think of some cases that we didn't cover? Hint: ACCOUNT-OWNER

Try to write a test that validates that only the correct owner of an account can vote.
:::

>  Notice the `let` construct which is helpful when you need to bind some variables to be in the same scope as some other logic that uses them. In our case we first loaded the number of votes and binded the result to `count` variable which we compared with the new count after submitting a vote. You can read more about `let` and `let*` <a href="https://pact-language.readthedocs.io/en/latest/pact-reference.html#let"> here</a>.

Reviewer: I think we should avoid using `here` as an anchor link, if for SEO purposes alone. It's better that the anchor links have descriptive text. For example:

```suggestion
 Notice the `let` construct which is helpful when you need to bind some variables to be in the same scope as some other logic that uses them. In our case we first loaded the number of votes and binded the result to `count` variable which we compared with the new count after submitting a vote. You can <a href="https://pact-language.readthedocs.io/en/latest/pact-reference.html#let">read more about `let` and `let*` in the Pact reference</a>.
```

Run those tests again to make sure everything works as expected:

```clojure
$ pact
pact> (load "election.repl")
```

:::info
The REPL preserves state between subsequent runs unless the optional parameter `reset` is set to true `(load "election.repl" true)`.
:::

> 

Reviewer: I feel like we should have a closing paragraph before we move on to the next section and start talking about gas stations. Just a quick summary of our tests, whatever we think would be helpful for the reader to remember when they're writing their own tests later. High-level.

### Gas Station

A unique feature of Kadena is the ability to allow gas to be paid by a different entity than the one who initiated the transaction. This entity is what we call a *gas station*.

>  In our voting app this will allow users to submit votes without paying for gas, instead gas will be subsidized by the gas station.

Reviewer: Maybe add a sentence: "In short, this means that miners will still be paid, but our users can vote for free."

The standard for gas station implementation is defined by the `gas-payer-v1` interface. The `gas-payer-v1` interface is deployed to all chains on `testnet` and `mainnet` so you can directly use it in your contract. We can specify that a module implements an interface using the `(implements INTERFACE)` construct.

:::info

Pact interfaces are similar to Java's interfaces, Scala's traits, Haskell's typeclasses or Solidity's interfaces.

> If you're not familiar with this concept you can read more about it <a href="https://pact-language.readthedocs.io/en/latest/pact-reference.html#interfaces">**here**</a>.

Reviewer: Same comment wrt `here` as an anchor link

:::

```clojure
(interface gas-payer-v1

  (defcap GAS_PAYER:bool
    ( user:string
      limit:integer
      price:decimal
    )

>     @doc

Reviewer: What is `@doc`? First time it's shown up in this file, and it would be useful to briefly explain it or link to the Pact reference.

    " Provide a capability indicating that declaring module supports \
    \ gas payment for USER for gas LIMIT and PRICE. Functionality \
    \ should require capability (coin.FUND_TX), and should validate \
    \ the spend of (limit * price), possibly updating some database \
    \ entry. \
    \ Should compose capability required for 'create-gas-payer-guard'."

>     @model

Reviewer: What is `@model`? First time it's shown up in this file, and it would be useful to briefly explain it or link to the Pact reference.

    [ (property (user != ""))
      (property (limit > 0))
      (property (price > 0.0))
    ]
  )

  (defun create-gas-payer-guard:guard ()
    @doc

>     " Provide a guard suitable for controlling a coin account that can \

Reviewer: Pact doesn't have multiline strings?

    \ pay gas via GAS_PAYER mechanics. Generally this is accomplished \
    \ by having GAS_PAYER compose an unparameterized, unmanaged capability \
    \ that is required in this guard. Thus, if coin contract is able to \
    \ successfully acquire GAS_PAYER, the composed 'anonymous' cap required \
    \ here will be in scope, and gas buy will succeed."
  )
)
```

Our module needs to implement all the functions and capabilities defined by the `gas-payer-v1` interface:
* `GAS_PAYER` capability
* `create-gas-payer-guard` function

A gas station allows someone to debit from a coin account that they do not own, gas station account, to pay the gas fee for a transaction under certain conditions. How exactly that happens, let's see below.

Create a new file `election-gas-station.pact` and paste the following snippet:

```clojure
(module election-gas-station GOVERNANCE
  (defcap GOVERNANCE ()
    "Only admin account can update the smart contract"
    (enforce-keyset 'election-admin-keyset))

  ; Implement the gas-payer-v1 interface
  (implements gas-payer-v1)

  ; Import the coin module
  (use coin)
)
```

> Now let's implement the `gas-payer-v1` interface:

Reviewer: This code snippet below definitely needs more explanation. Where is `chain-data` coming from? What is it? What does it mean to be an `exec` transaction? Why do we test that the length is 1 at `exec-code`? Where did the `free-election.` module come from? Why are we taking 15 from the exec-code? What does it mean to compose capabilities (ie. the call to `compose-capability`)?

I feel like a lot of code in this snippet below requires introduction and an explanation as to what it's doing and why.

---
title: dApp Tutorial
description: Build a complete dApp on Kadena
---

import PageRef from '@components/PageRef'
import Tabs from '@theme/Tabs';
import TabItem from '@theme/TabItem';

> # Complete dApp Tutorial

Reviewer 2: Can we come up with a more descriptive title? Maybe something like: `First steps with Kadena` or something catchier

Reviewer 2: (adding this comment in the hopes that the above gets added to my "review")

Author: From "zero to hero" with Kadena?

> # Complete dApp Tutorial

Reviewer 3: Building a voting dApp

Author: Jacquin also left a comment here to come up with a catchier title: https://github.com/raduciobanu22/docs/pull/1#discussion_r837768134
Let's move the conversation over there, there's 2 suggestions for now.

> Build a complete on-chain voting dApp

Reviewer 2: What does "on-chain voting dapp" mean? This is going to be THE FIRST thing that new developers see when they begin learning about us. Let's try and make sure that the tutorials objectives are _super_ clear

Author: I integrated the paragraph that Bertrand provided

> Build a complete on-chain voting dApp

Reviewer 3: In its most basic form, the Kadena blockchain is a digital ledger. This ledger is never stored, but rather exists on the “chain” supported by thousands of nodes simultaneously. Thanks to encryption and decentralization, the Kadena blockchain’s database of transactions is incorruptible, and each record is easily verifiable. The network cannot be taken down or influenced by a single party because it doesn’t exist in one place.

Blockchain can solve the many problems discovered in these early attempts at online voting. A blockchain-based voting application does not concern itself with the security of its Internet connection, because any hacker with access to the terminal will not be able to affect other nodes. Voters can effectively submit their vote without revealing their identity or political preferences to the public. Officials can count votes with absolute certainty, knowing that each ID can be attributed to one vote, no fakes can be created, and that tampering is impossible.

Author: Thanks

---

## Overview

> In this tutorial you will build a complete dApp, smart contracts and frontend that allows users to submit a vote. You will also deploy your own gas station that will be used to pay for the gas needed to interact with your dApp.

Reviewer 3: In this tutorial we will be building a voting application on the Kadena blockchain. The application will consist of three distinct layers:

1. Smart contract (_back-end_)
2. Web app (_front-end_)
3. [Gas Station](https://medium.com/kadena-io/users-shouldnt-pay-for-gas-4df989ec8236) (_account owned by the voting application funding gas payments on behalf of the voters_)

Author: Is that Medium article the best one we have? It's from 2019 and Will presents the idea as a concept while today it's reality.

Author: How about this one: https://medium.com/kadena-io/the-first-crypto-gas-station-is-now-on-kadenas-blockchain-6dc43b4b3836

> By leveraging the blockchain and smart contracts to implement this application, its users benefit of complete transparency of the voting process, anyone being able to check every vote that was recorded, as well as peace of mind that their vote will never be deleted. This set of features is simply not possible in a classic Web2 application.

Reviewer 3: Delete. Content has been moved raw 12.

## Environment Setup

### Install Pact

<Tabs>
  <TabItem value="mac" label="Mac">

    brew update
    brew install kadena-io/pact/pact
  </TabItem>

  <TabItem value="linux" label="Linux">

>     For installing Pact on Linux distributions in the Arch family, refer to <a href="https://aur.archlinux.org/packages/pact/">this package on the AUR</a>. Otherwise, please refer to <a href="https://github.com/kadena-io/pact#building-from-source">building from source</a>.

Reviewer 2: We haven't updated that AUR package in forever. Let's maybe copy the walk through of installing via `nix` instead (and also link to the `buildFromSrc`

Reviewer 2: (This comment should make the above visible in the official "review")

Author: Updated. And I just found out there's a Wiki.

  </TabItem>
</Tabs>

**Verifying Installation**

Type `pact` in a terminal and try out some commands:

```clojure
$ pact
pact> (+ 1 2)
3
pact> (+ "hello, " "world")
"hello, world"
pact> (< 0 1)
true
```

### Create Project Structure

> Copy paste the instructions below to create a basic project structure:

Reviewer 2: Remove this statement, it's redundant given the title

Author: Removed

```bash
mkdir my-pact-project && cd my-pact-project
mkdir pact
mkdir app
```

### Atom IDE (Optional)

You can use any text editor to write Pact but if you prefer the benefits of an IDE, "language-pact" is a package for [Atom](https://atom.io) that provides syntax highlighting and linting.

> To install it, go to Preferences -> Packages -> Type "language-pact" and click the Install button.

Reviewer 2: "Preferences -> Packages -> Type "language-pact"" : make this an inline code snippet

![atom-language-pact](/img/docs/voting-dapp/atom-language-pact.png)

## Smart Contracts

> Our voting app should allow you to submit a vote while not allowing an address to vote more than once.

Reviewer 2: "...app should allow" --> "will allow"

Author: Fixed

Additionally we will use a *gas station* to fund the gas fees for interacting with our contract, meaning our users don't have to worry about paying gas.

:::info
**Gas** is the cost necessary to perform a transaction on the network. Gas is paid to miners and its price varies based on supply and demand. It's a critical piece of the puzzle, but at the same time it brings up a UX problem. Every user needs to be aware of what gas is as well as how much gas they need to pay for their transaction. This causes significant friction and a less than ideal experience.

> To help mitigate this problem Kadena brings an innovation to the game. Hello **gas stations**! Gas stations are a way for dApps to subsidize gas costs for their users. This means that your user doesn't need to know what gas is or how much the gas price is, which translates into a smooth experience when interacting with your dApp.

Reviewer 2: Let's link to this as well: https://medium.com/kadena-io/the-first-crypto-gas-station-is-now-on-kadenas-blockchain-6dc43b4b3836

Reviewer 2: Let's also make a note to migrate/inline this article to the docs on its own separate page

Author: Linked and left a comment in the excel doc.

:::

### Voting

> Let's start to implement our voting contract.

Reviewer 2: Redundant statement

A typical developer workflow looks like this:

> 1. Write logic in `.pact` files

Reviewer 2: *write contract code

2. Write tests in `.repl` files
3. Execute your tests in the repl
4. Deploy to local pact server
5. Deploy to testnet

Here we will be focusing on steps 1-3.

Assuming you are still in your project directory, create the `election.pact` and `election.repl` files:

```clojure
touch pact/election.pact pact/election.repl
```

`election.pact` is the contract source code while in `election.repl` we'll write tests.

> We will also need the `coin` contract which requires the `fungible-v2` interface for our tests. You can get the latest `coin-v3` [here](https://github.com/kadena-io/chainweb-node/blob/master/pact/coin-contract/v3/coin-v3.pact) and `fungible-v2` [here](https://github.com/kadena-io/chainweb-node/blob/master/pact/coin-contract/v2/fungible-v2.pact). Make sure to add those files in your project.

Reviewer 2: Best practice dictates that these need to be placed in `pact/root/` (see structure here: https://github.com/kadena-io/marmalade/tree/main/pact/root)

Author: Moved them in the source code repo as well

Now copy the following code in the `election.pact` file:

```clojure
(define-keyset 'election-admin-keyset)

(module election GOVERNANCE
  "Election demo module"

  (defcap GOVERNANCE ()
    "Only admin can update this module"
    (enforce-keyset 'election-admin-keyset))

  (defun vote (key)
    "Submit a new vote"
    (format "Voted {}!" [key]))
)

(vote "A")
```

This snippet defines a module "election" that holds a function "vote" that takes a parameter and displays a formatted string that includes the parameter.

:::note
Module names and keyset definitions are required to be unique. We will mention this again when we get to deploy our contract to testnet.
:::

Before we move forward, let's quickly test our code. Copy the code below in the `election.repl` file:

```clojure
;; begin a transaction
(begin-tx)
;; set transaction signature key to my-key
(env-keys ["my-key"])
;; Add the election-admin-keyset to environment data
(env-data {
  'election-admin-keyset: { "keys": ["admin-key"], "pred": "keys-all" }  
})
;; load election.pact into the REPL
(load "election.pact")
;; commit the transaction
(commit-tx)
```

> 

Reviewer 2: Add another ::info or ::note that explains that each pact transaction is submitted with raw json data that functions can read from, and where we can store our keysets

Author: I added it and mentioned `read-msg` and `read-keyset` functions,  hope it's what you had in mind.

```clojure
"Loading election.repl..."
"Begin Tx 0"
"Setting transaction keys"
"Setting transaction data"
"Loading election.pact..."
"Keyset defined"
"Loaded module election, hash jOBJiSEH5HVsAmLNSdHq_D3Kl6TkFpCJcEWPtJD86Gc"
"Voted A!"
"Commit Tx 0"
```

Let's continue by adding a bit more logic. Copy this snippet just below the module declaration:

```clojure
  (defschema candidates-schema
    "Candidates table schema"
    name:string
    votes:integer)

  (defschema votes-schema
    "Votes table schema"
    cid:string
  )

  (deftable votes:{votes-schema})

  (deftable candidates:{candidates-schema})
  ```
In the snippet we've got several definitions:
* `defschema candidates-schema` -> defines the schema for our **candidates** table where we will keep track of the candidates that users can vote for
* `defschema votes-schema` -> defines the schema for our **votes** table where we will store each vote
* `deftable votes:{votes-schema}` -> defines the `votes` table that will use the schema defined above
* `deftable candidates:{candidates-schema}` -> defines the `candidates` table that will use the schema defined above

:::note

> Pact implements a key-row model which means a row is accessed by a single key. It is our responsibility as developers to design the schema in a way that we can retrieve the information that we need using a single row query. Multiple row queries are expensive which we will see in an example a bit later.

Reviewer 2: Make a note that the key is always a string. Maybe also note that "key" here is not the same as a "cryptographic key" that we use for signing? I wonder if that might get confusing

Author: Hmm, it should not be for anyone that worked with key-value dbs before. I left note though just in case.

:::

Let's continue by implementing our `vote` function:

```clojure
  (defun vote (account:string cid:string)
    "Submit a new vote"

>     (let ((doubleVote (user-voted account)))

Reviewer 2: `double-vote` not `doubleVote` -- Needs to be consistent

    (let ((exists (candidate-exists cid)))
      (enforce (= exists true) "Candidate doesn't exist"))

    (with-capability (ACCOUNT-OWNER account)
      (vote-protected account cid))

    (format "Voted for candidate {}!" [cid])
  )
```

Quite a lot happening here, so let's go line by line:
* to prevent double voting, we have to check if the user already voted. Below is the implementation of the `user-voted` function which does that.
```clojure
(defun user-voted:bool (account:string)
  (with-default-read votes account
    { "cid": "" }
    { "cid":= cid }
    (> (length cid) 0))
)
```
* to prevent voting for a candidate that doesn't exist we implemented `candidate-exists` function and we `enforce` it returns `true` for our parameters

```clojure
(defun candidate-exists:bool (cid:string)
  (with-default-read candidates cid
    { "name": "" }
    { "name" := name }
    (> (length name) 0))
)
```

:::note
Notice how we used a similar approach in both `user-voted` and `candidate-exists` functions to execute a single row read using `with-default-read`. Below is an alternative, naive a way to implement the same thing:

```clojure

> (contains account (keys votes))

Reviewer 2: Would rather not even mention `keys` here. Just pretend like it doesn't exist

Author: My opinion here is that is very likely people will use `keys` and "how not to do it" example is useful. But then yes, you have them wondering what's the purpose of `keys`.

Author: I removed that example. I fear that many will still use this function because it's in the docs.

```clojure
; import coin.details function
(use coin [ details ])

(defcap ACCOUNT-OWNER (account:string)
  "Make sure the requester owns the KDA account"
  (enforce-guard (at 'guard (coin.details account)))
)
```
* while the `ACCOUNT-OWNER` capability is in scope we are calling `vote-protected` which is the function that updates the database
```clojure
(defun vote-protected (account:string candidateId:string)
  (require-capability (ACCOUNT-OWNER account))

  (with-read candidates candidateId { "votes" := votesCount }
    (update candidates candidateId { "votes": (+ votesCount 1) })
    (insert votes account { "cid": candidateId })
    (emit-event (VOTED candidateId))
  )
)
```
First it checks that we have the capability, then proceed to update the votes of a candidate and record the vote in the votes table. At the end we emit the `VOTED` event which is useful for our frontend to update itself in real-time by listening to this event.

:::tip
Spend a bit of time understanding the logic in the `vote` function before moving forward.
:::

We're missing a way to read the current number of votes so here it is:

```clojure

> (defun getVotes:integer (cid:string)

Reviewer 2: `get-votes` not `getVotes`

  "Get the votes count by cid"
  (at 'votes (read candidates cid ['votes]))
)
```

And we also need to add some candidates in our table:

```clojure
(defun init ()
  "Initialize the rows in candidates table"

>   (insert candidates "1" { "name": "Candidate A", "votes": 0 })

Reviewer 2: Can we add an `insert-candidate` function that can only be performed by admin? Perhaps it takes a list so that it can be done in bulk? Would be a good way to demonstrate govt / guarding code

---
title: dApp Tutorial
description: Build a complete dApp on Kadena
---

import PageRef from '@components/PageRef'
import Tabs from '@theme/Tabs';
import TabItem from '@theme/TabItem';

# Building a voting dApp

> In its most basic form, Kadena blockchain is a digital ledger. This ledger is never stored, but rather exists on the “chain” supported by thousands of nodes simultaneously. Thanks to encryption and decentralization, Kadena blockchain’s database of transactions is incorruptible, and each record is easily verifiable. The network cannot be taken down or influenced by a single party because it doesn’t exist in one place.

Reviewer 4: I think there needs to be an introductory section to the tutorial. The current introduction in `Building a voting dApp` explains why the example app being built is important, which is good to include. But I think if someone is coming to this tutorial, they might not care too much about the app itself, and instead they might care more about what the different components of this end-to-end tutorial will be and why each component is important.

Author: I think you're right about not caring much about the app itself but learning the process of building **any app** on Kadena. The process and its components are the take aways from this tutorial.

Blockchain can solve the many problems discovered in these early attempts at online voting. A blockchain-based voting application does not concern itself with the security of its Internet connection, because any hacker with access to the terminal will not be able to affect other nodes. Voters can effectively submit their vote without revealing their identity or political preferences to the public. Officials can count votes with absolute certainty, knowing that each ID can be attributed to one vote, no fakes can be created, and that tampering is impossible.

---

## Overview

In this tutorial we will be building a voting application on the Kadena blockchain. The application will consist of three distinct layers:

1. Smart contract (back-end)
2. Web app (front-end)

> 3. Gas Station

Reviewer 4: I did a quick pass of the entire tutorial and, at first, I thought this section would be a workflow of how the tutorial would be organized. 

But the tutorial's main sections were actually:
`Environment Setup -> Smart Contracts -> Testing -> Deploy Contract -> Frontend -> Conclusion`

I think one approach would be to include a summary of what the big sections of the tutorial will be, and/or a Table of Contents with links to the different sections and sub-sections.

---
title: dApp Tutorial
description: Build a complete dApp on Kadena
---

import PageRef from '@components/PageRef'
import Tabs from '@theme/Tabs';
import TabItem from '@theme/TabItem';

# Complete dApp Tutorial

Build a complete on-chain voting dApp

---

## Overview

> In this tutorial you will build a complete dApp, smart contracts and frontend that allows users to submit a vote. You will also deploy your own gas station that will be used to pay for the gas needed to interact with your dApp.

Reviewer 2: Lets try and add a few paragraphs to motivate why a decentralized ledger is a good fit for this sort of application / what a smart contract is able to do that would otherwise not be possible on web2-style tech

Author: Added a paragraph here, feedback is welcome.

Reviewer 2: "to submit a vote" -- Can we elaborate ? Ex: "....that allow users to record a vote between two parties -- set by the contract administrator -- in a completely tamper-proof/censorship-resistant fashion"

Author: I also thought "submit a vote" sounds pretty lame, thanks for the suggestion

## Environment Setup

### Install Pact

> **Using Brew**

Reviewer 2: Lets take this whole section and either A) Make it part of the cookbook and link to it, or B) At least have a tabbed window that shows how to do it on all OSs

Author: I went with option B.

```clojure
brew update
brew install kadena-io/pact/pact
```
**Build from source**
- follow the instructions [here](https://github.com/kadena-io/pact#building-from-source)

**Verifying Installation**

Type `pact` in a terminal and try out some commands:

```clojure
$ pact
pact> (+ 1 2)
3
pact> (+ "hello, " "world")
"hello, world"
pact> (< 0 1)
true
```

### Create Project Structure

Copy paste the instructions below to create a basic project structure:

```clojure
mkdir my-pact-project && cd my-pact-project
mkdir pact
mkdir app
```

### Pact Http Server

Create a `config.yaml` file for pact http server. We will later use it simulate blockchains transactions locally.

For more details about pact http server check <a href="/blog/deploy-to-a-local-server">this</a> tutorial.

```clojure
touch config.yaml
```
Copy the code below and save the file:

```clojure
# Config file for pact http server. Launch with `pact -s config.yaml`

# HTTP server port
port: 8080

# directory for HTTP logs

> logDir: log

Reviewer 2: Need to `mkdir log` too or else starting the server doesn't work

Author: I moved the Pact Server setup to the Pact Server section since that's where it's used and added `mkdir`.

# persistence directory
persistDir: log

# SQLite pragmas for pact back-end
pragmas: []

# verbose: provide log output
verbose: True
```

### Atom IDE (Optional)

You can use any text editor to write Pact but if you prefer the benefits of an IDE, "language-pact" is a package for [Atom](https://atom.io) that provides syntax highlighting and linting.

To install it, go to Preferences -> Packages -> Type "language-pact" and click the Install button.

![atom-language-pact](/img/docs/voting-dapp/atom-language-pact.png)

## Smart Contracts

Our voting app should allow you to submit a vote while not allowing an address to vote more than once.

> Additionally we will use a *gas station* to fund the gas fees for interacting with our contract, meaning our users don't have to worry about paying gas.

Reviewer 2: Elaborate more on what gas is, what a gas station is, and why they are so important

Author: Explained gas, the problem and gas stations as solution to the problem.

Reviewer 5: I do think gas stations etc should not be part of this doc but we should link to a page where we explain this.

### Voting

Let's start to implement our voting contract.

Assuming you are still in your project directory, create the `vote.pact` and `vote.repl` files:

```clojure
touch pact/vote.pact pact/vote.repl
```

> `vote.pact` is the contract source code while in `vote.repl` we'll write tests.

Reviewer 2: Explain a typical dev workflow here

Author: Added it, let me know if you think we should go into more details.

Copy the following code in the `vote.pact` file:

```clojure
(define-keyset 'vote-admin-keyset)

> (module simple-vote 'vote-admin-keyset

Reviewer 2: Have one of those "Warning boxes" here that points out that module names need to be unique, and as such, they will need to choose a different module name when they go to actually deploy to testnet. Also maybe use the free namespace from the start?

Author: I added it and I also mention this in the deployment section. Regarding the namespace, I mentioned it at the deployment step cause otherwise I'd have to say that you need to comment it while you're running tests in REPL or I need to write more code for that namespace definition for testing purposes and I thought it's too much overload at start.

Reviewer 5: Rules like this should be in some sort of pact rules part where we again ca link towards for now I think this is pretty nice

  "Simple voting module"

  (defun vote (key)
    "Submit a new vote"
    (format "Voted {}!" [key]))
)

(vote "A")
```

This snippet defines a module "simple-vote" that holds a function "vote" that takes a parameter and displays a formatted string that includes the parameter.

Before we move forward, let's quickly test our code. Copy the code below in the `vote.repl` file:

```clojure
;; begin a transaction
(begin-tx)
;; set transaction signature key to my-key

> (env-keys ["my-key"])

Reviewer 2: Explaiin the notion of "repl-only" functions, and explain that these functions are simulating the api structures that are represented by the chainweb-blockchain

Author: Explained it.

```
And you should see an output similar to the one below (the hash will be different):
```clojure
"Loading vote.repl..."
"Begin Tx 0"
"Setting transaction keys"
"Setting transaction data"
"Loading vote.pact..."
"Keyset defined"
"Loaded module simple-vote, hash jOBJiSEH5HVsAmLNSdHq_D3Kl6TkFpCJcEWPtJD86Gc"
"Voted A!"
"Commit Tx 0"
```

Let's continue by adding a bit more logic. Copy this snippet just below the module declaration:

```clojure
  (defschema votes-schema
    "Votes table schema"
    option:string
    count:integer)

  (deftable votes:{votes-schema})
  ```
In the snippet we've got 2 new declaration:
* `defschema votes-schema` -> defines the schema for our table where we will keep track of the number of votes
* `deftable votes:{votes-schema}` -> defines the `votes` table that will use the schema defined above

Now we need to update our `vote` function:

```clojure
(defun vote (key:string)
    "Submit a new vote"
    (enforce (contains key ["A" "B"]) "You can only vote for A or B")
    (with-read votes key { "count" := count }
      (update votes key { "count": (+ count 1) })
    )
    (format "Voted {}!" [key])
  )
```

Again, quick recap of what's happening above:

> * we use `enforce` to ensure that only the given list of vote keys are allowed

Reviewer 2: can we rephrase this ? Or better yet, come up with a better way than submitting raw strings as votes

Author: This was refactored.

    (at 'count (read votes key ['count]))
)
```

And we also need a way to initialize our table:

```clojure
(defun init ()
    "Initialize the rows in votes table"
    (insert votes "A" { "option": "A", "count" : 0 })
    (insert votes "B" { "option": "B", "count" : 0 })
)
```

If you followed the steps correctly, your code should look similar to this:
```clojure
(define-keyset 'vote-admin-keyset)

(module simple-vote 'vote-admin-keyset
  "Simple voting module"

  (defschema votes-schema
    "Votes table schema"
    option:string
    count:integer)

  (deftable votes:{votes-schema})

  (defun vote (key:string)
    "Submit a new vote"
    (enforce (contains key ["A" "B"]) "You can only vote for A or B")
    (with-read votes key { "count" := count }
      (update votes key { "count": (+ count 1) })
    )
    (format "Voted {}!" [key])
  )

  (defun getVotes:integer (key:string)
    "Get the votes count by key"
    (at 'count (read votes key ['count]))
  )

  (defun init ()
    "Initialize the rows in votes table"
    (insert votes "A" { "option": "A", "count" : 0 })
    (insert votes "B" { "option": "B", "count" : 0 })
  )
)

(create-table votes)
(init)
```

> Notice how we're calling `create-table` and `init` outside of our module. That's an important step, so don't miss it.

Reviewer 2: Explain how code outside module executes one on run

Author: Added a note about this.

```
Notice how we are using `expect` to test if `getVotes` returns the correct result.
`expect` is one the functions that's only available in the REPL but there are many more, check them out <a href="https://pact-language.readthedocs.io/en/latest/pact-functions.html#repl-only-functions">here</a>.

I also want to mention the `let` construct which is helpful when you need to bind some variables be in the same scope as some other logic that makes use of them. In our case we first loaded the number of votes and binded the result to `count` variable which we used later in the `expect` function.

You can read more about `let` and `let*` <a href="https://pact-language.readthedocs.io/en/latest/pact-reference.html#let"> here</a>.

Let's run those tests again to make sure everything works as expected:

```clojure
$ pact
pact> (load "vote.repl")
```

### Gas Station

A unique feature of Kadena is the ability to allow gas to be paid by a different entity than the one who initiated the transaction. This entity is what we call a "gas station".

Let's create our own gas station that will allow users to submit votes without paying for gas, instead gas will be subsidized by the gas station.

> Each gas station needs to implement the `gas-payer-v1` interface which is shown below:

Reviewer 2: Use this opportunity to explain what an interface is, and that you can find this particular one on chain X (all of them? )

Author: There's an info section just below the snippet that explains it.

Reviewer 5: IMHO this shouldn't be part of this documentation but part of another section that we can link to. We implement a gas station so we can tell why we create a gas station but not what a gas station is and does. I do like the text we should just move it.
